File: 0000-0999/0126.WordLadderII.py
import collections
import string

class Solution(object):
    def findLadders(self, beginWord, endWord, wordlist):
        """
        :type beginWord: str
        :type endWord: str
        :type wordlist: Set[str]
        :rtype: List[List[int]]
        """
        wordlist.add(endWord)
        wordLen = len(beginWord)
        level = {beginWord}
        parents = collections.defaultdict(set)
        while level and endWord not in parents:
            nextLevel = collections.defaultdict(set)
            for word in level:
                for ch in string.ascii_lowercase:
                    for i in xrange(wordLen):
                        temp = word[:i] + ch + word[i+1:]
                        if temp in wordlist and temp not in parents:
                            nextLevel[temp].add(word)
            level = nextLevel
            parents.update(nextLevel)
        res = [[endWord]]
        while res and res[0][0] != beginWord:
            res = [[p]+r for r in res for p in parents[r[0]]]
        return res
        '''
        dic.add(end)
        level = {start}
        parents = collections.defaultdict(set)
        while level and end not in parents:
            next_level = collections.defaultdict(set)
            for node in level:
                for char in string.ascii_lowercase:
                    for i in range(len(start)):
                        n = node[:i]+char+node[i+1:]
                        if n in dic and n not in parents:
                            next_level[n].add(node)
            level = next_level
            parents.update(next_level)
        res = [[end]]
        while res and res[0][0] != start:
            res = [[p]+r for r in res for p in parents[r[0]]]
        return res
        '''

File: 0000-0999/0126.WordLadderII.go
func findLadders(beginWord string, endWord string, wordList []string) [][]string {
    Map := make(map[string]bool, len(wordList))
    for _, word := range wordList {
        Map[word] = true
    }
    //Map[beginWord] = false
    ans := [][]string{}
    step, minStep := 1, math.MaxInt32
    paths := [][]string{[]string{beginWord}}
    visited := []string{}
    for len(paths) > 0 {
        path := paths[0]
        paths = paths[1:]
        if len(path) > step {
            if len(path) > minStep {
                break
            } else {
                step = len(path)
            }
            for _, word := range visited {
                Map[word] = false
            }
            visited = []string{}
        }
        cur := path[len(path)-1]
        for i := 0; i < len(cur); i++ {
            for ch := 'a'; ch <= 'z'; ch++ {
                nxt := cur[:i] + string(ch) + cur[i+1:]
                if Map[nxt] {
                    newPath := make([]string, len(path)+1)
                    copy(newPath, path)
                    newPath[len(path)] = nxt
                    visited = append(visited, nxt)
                    if nxt == endWord {
                        minStep = step
                        ans = append(ans, newPath)
                    } else {
                        paths = append(paths, newPath)
                    }
                }
            }
        }
    }
    return ans
}